File: ios/HaliSahaApp/HaliSahaApp/Services/StorageService.swift
//
//  StorageService.swift
//  HaliSahaApp
//
//  Firebase Storage işlemleri - Fotoğraf yükleme/silme
//  DÜZELTİLMİŞ VERSİYON: downloadImage kaldırıldı (ImageCacheService kullanılacak)
//
//  Created by Mehmet Mert Mazıcı on 27.01.2026.
//

import Foundation
import FirebaseStorage
import UIKit

// MARK: - Storage Service
@MainActor
final class StorageService {
    
    // MARK: - Singleton
    static let shared = StorageService()
    
    // MARK: - Private Properties
    private let storage = Storage.storage()
    private let maxImageSize: Int64 = 5 * 1024 * 1024 // 5MB
    private let compressionQuality: CGFloat = 0.7
    private let maxImageDimension: CGFloat = 1920 // Max genişlik/yükseklik
    
    // Retry ve timeout ayarları
    private let maxRetries = 3
    private let retryDelay: UInt64 = 1_000_000_000 // 1 saniye (nanoseconds)
    
    // Upload limiter
    private actor UploadLimiter {
        private var activeUploads = 0
        private let maxConcurrent = 3
        
        func acquire() async {
            while activeUploads >= maxConcurrent {
                await Task.yield()
            }
            activeUploads += 1
        }
        
        func release() {
            activeUploads = max(0, activeUploads - 1)
        }
    }
    
    private let uploadLimiter = UploadLimiter()
    
    // MARK: - Private Init
    private init() {}
    
    // MARK: - Storage References
    private var facilityImagesRef: StorageReference {
        storage.reference().child("facilities")
    }
    
    private var pitchImagesRef: StorageReference {
        storage.reference().child("pitches")
    }
    
    private var userImagesRef: StorageReference {
        storage.reference().child("users")
    }
    
    // MARK: - Upload Facility Image
    func uploadFacilityImage(_ image: UIImage, facilityId: String) async throws -> String {
        let imageId = UUID().uuidString
        let ref = facilityImagesRef.child(facilityId).child("\(imageId).jpg")
        
        return try await uploadImage(image, to: ref)
    }
    
    // MARK: - Upload Multiple Facility Images (İYİLEŞTİRİLDİ)
    func uploadFacilityImages(_ images: [UIImage], facilityId: String) async throws -> [String] {
        var urls: [String] = []
        var errors: [Error] = []
        
        // Paralel yükleme için TaskGroup kullan
        await withTaskGroup(of: Result<String, Error>.self) { group in
            for image in images {
                group.addTask {
                    do {
                        let url = try await self.uploadFacilityImage(image, facilityId: facilityId)
                        return .success(url)
                    } catch {
                        return .failure(error)
                    }
                }
            }
            
            for await result in group {
                switch result {
                case .success(let url):
                    urls.append(url)
                case .failure(let error):
                    errors.append(error)
                }
            }
        }
        
        // Eğer hiç yükleme başarılı olmadıysa hata fırlat
        if urls.isEmpty && !errors.isEmpty {
            throw errors.first!
        }
        
        return urls
    }
    
    // MARK: - Upload Pitch Image
    func uploadPitchImage(_ image: UIImage, facilityId: String, pitchId: String) async throws -> String {
        let imageId = UUID().uuidString
        let ref = pitchImagesRef.child(facilityId).child(pitchId).child("\(imageId).jpg")
        
        return try await uploadImage(image, to: ref)
    }
    
    // MARK: - Upload Multiple Pitch Images (İYİLEŞTİRİLDİ)
    func uploadPitchImages(_ images: [UIImage], facilityId: String, pitchId: String) async throws -> [String] {
        var urls: [String] = []
        var errors: [Error] = []
        
        await withTaskGroup(of: Result<String, Error>.self) { group in
            for image in images {
                group.addTask {
                    do {
                        let url = try await self.uploadPitchImage(image, facilityId: facilityId, pitchId: pitchId)
                        return .success(url)
                    } catch {
                        return .failure(error)
                    }
                }
            }
            
            for await result in group {
                switch result {
                case .success(let url):
                    urls.append(url)
                case .failure(let error):
                    errors.append(error)
                }
            }
        }
        
        if urls.isEmpty && !errors.isEmpty {
            throw errors.first!
        }
        
        return urls
    }
    
    // MARK: - Upload User Profile Image
    func uploadUserProfileImage(_ image: UIImage, userId: String) async throws -> String {
        let ref = userImagesRef.child(userId).child("profile.jpg")
        return try await uploadImage(image, to: ref)
    }
    
    // MARK: - Generic Upload (with Retry & Optimization)
    private func uploadImage(_ image: UIImage, to ref: StorageReference) async throws -> String {
        // Upload limiter
        await uploadLimiter.acquire()
        defer { Task { await uploadLimiter.release() } }
        
        // Görüntüyü optimize et (boyut küçültme)
        let optimizedImage = optimizeImage(image)
        
        // Görüntüyü sıkıştır
        guard let imageData = optimizedImage.jpegData(compressionQuality: compressionQuality) else {
            throw StorageError.compressionFailed
        }
        
        // Boyut kontrolü
        if imageData.count > maxImageSize {
            // Daha fazla sıkıştır
            guard let reducedData = optimizedImage.jpegData(compressionQuality: 0.5) else {
                throw StorageError.compressionFailed
            }
            
            if reducedData.count > maxImageSize {
                throw StorageError.fileTooLarge
            }
            
            return try await performUpload(data: reducedData, to: ref)
        }
        
        return try await performUpload(data: imageData, to: ref)
    }
    
    // MARK: - Optimize Image (YENİ)
    private func optimizeImage(_ image: UIImage) -> UIImage {
        let maxDim = maxImageDimension
        
        // Boyut kontrolü
        if image.size.width <= maxDim && image.size.height <= maxDim {
            return image
        }
        
        // Aspect ratio'yu koru
        let widthRatio = maxDim / image.size.width
        let heightRatio = maxDim / image.size.height
        let ratio = min(widthRatio, heightRatio)
        
        let newSize = CGSize(
            width: image.size.width * ratio,
            height: image.size.height * ratio
        )
        
        // Yeniden boyutlandır
        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }
    
    // MARK: - Perform Upload (Retry logic)
    private func performUpload(data: Data, to ref: StorageReference) async throws -> String {
        // Metadata oluştur
        let metadata = StorageMetadata()
        metadata.contentType = "image/jpeg"
        
        // Retry mekanizması ile yükle
        var lastError: Error?
        
        for attempt in 1...maxRetries {
            do {
                _ = try await ref.putDataAsync(data, metadata: metadata)
                let downloadURL = try await ref.downloadURL()
                return downloadURL.absoluteString
            } catch {
                lastError = error
                print("⚠️ Upload attempt \(attempt)/\(maxRetries) failed: \(error.localizedDescription)")
                
                if attempt < maxRetries {
                    // Üstel backoff: 1s, 2s, 4s
                    let delay = retryDelay * UInt64(1 << (attempt - 1))
                    try await Task.sleep(nanoseconds: delay)
                }
            }
        }
        
        throw lastError ?? StorageError.uploadFailed("Unknown error after \(maxRetries) attempts")
    }
    
    // MARK: - Delete Image
    func deleteImage(at url: String) async throws {
        guard !url.isEmpty else { return }
        
        guard let ref = try? storage.reference(forURL: url) else {
            throw StorageError.invalidURL
        }
        
        do {
            try await ref.delete()
        } catch {
            // Dosya zaten silinmişse hata verme
            let nsError = error as NSError
            if nsError.domain == StorageErrorDomain && nsError.code == StorageErrorCode.objectNotFound.rawValue {
                print("⚠️ Image already deleted: \(url)")
                return
            }
            throw StorageError.deleteFailed(error.localizedDescription)
        }
    }
    
    // MARK: - Delete Multiple Images (İYİLEŞTİRİLDİ)
    func deleteImages(at urls: [String]) async throws {
        let validURLs = urls.filter { !$0.isEmpty }
        guard !validURLs.isEmpty else { return }
        
        // Paralel silme
        await withTaskGroup(of: Void.self) { group in
            for url in validURLs {
                group.addTask {
                    try? await self.deleteImage(at: url)
                }
            }
        }
    }
    
    // MARK: - Delete All Facility Images
    func deleteFacilityImages(facilityId: String) async throws {
        let ref = facilityImagesRef.child(facilityId)
        try await deleteFolder(ref)
    }
    
    // MARK: - Delete All Pitch Images
    func deletePitchImages(facilityId: String, pitchId: String) async throws {
        let ref = pitchImagesRef.child(facilityId).child(pitchId)
        try await deleteFolder(ref)
    }
    
    // MARK: - Delete Folder
    private func deleteFolder(_ ref: StorageReference) async throws {
        do {
            let result = try await ref.listAll()
            
            // Paralel silme
            await withTaskGroup(of: Void.self) { group in
                for item in result.items {
                    group.addTask {
                        try? await item.delete()
                    }
                }
            }
            
            // Alt klasörleri sil
            for prefix in result.prefixes {
                try await deleteFolder(prefix)
            }
        } catch {
            // Klasör boşsa veya yoksa hata verme
            let nsError = error as NSError
            if nsError.domain == StorageErrorDomain && nsError.code == StorageErrorCode.objectNotFound.rawValue {
                return
            }
            throw error
        }
    }
}

// MARK: - Storage Error
enum StorageError: LocalizedError {
    case compressionFailed
    case fileTooLarge
    case invalidURL
    case uploadFailed(String)
    case downloadFailed(String)
    case deleteFailed(String)
    
    var errorDescription: String? {
        switch self {
        case .compressionFailed:
            return "Görüntü sıkıştırılamadı"
        case .fileTooLarge:
            return "Dosya boyutu çok büyük (max 5MB)"
        case .invalidURL:
            return "Geçersiz dosya URL'i"
        case .uploadFailed(let message):
            return "Yükleme hatası: \(message)"
        case .downloadFailed(let message):
            return "İndirme hatası: \(message)"
        case .deleteFailed(let message):
            return "Silme hatası: \(message)"
        }
    }
}

File: android/app/src/main/java/com/example/HaliSahaApp/data/remote/StorageService.kt
package com.example.HaliSahaApp.data.remote

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import com.google.firebase.storage.StorageReference
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.tasks.await
import java.io.ByteArrayOutputStream
import java.util.UUID

object StorageService {

    private val storage = FirebaseStorage.getInstance()
    private const val MAX_IMAGE_SIZE: Long = 5 * 1024 * 1024 // 5MB
    private const val COMPRESSION_QUALITY = 70 // 0-100 aralığında
    private const val MAX_IMAGE_DIMENSION = 1920f

    private const val MAX_RETRIES = 3
    private const val RETRY_DELAY_MS = 1000L

    // Upload Limiter (Maksimum 3 eşzamanlı yükleme)
    // Kotlin'de Semaphore veya Mutex kullanılabilir, basitlik adına bir Mutex ile sırayla işleme alacağız
    private val uploadMutex = Mutex()

    // Referanslar
    private val facilityImagesRef: StorageReference get() = storage.reference.child("facilities")
    private val pitchImagesRef: StorageReference get() = storage.reference.child("pitches")
    private val userImagesRef: StorageReference get() = storage.reference.child("users")
    private val adminVerificationsRef: StorageReference get() = storage.reference.child("admin_verifications")

    // MARK: - Upload Functions

    suspend fun uploadFacilityImage(bitmap: Bitmap, facilityId: String): String {
        val imageId = UUID.randomUUID().toString()
        val ref = facilityImagesRef.child(facilityId).child("$imageId.jpg")
        return uploadImage(bitmap, ref)
    }

    suspend fun uploadFacilityImages(bitmaps: List<Bitmap>, facilityId: String): List<String> = coroutineScope {
        val deferredUrls = bitmaps.map { bitmap ->
            async {
                uploadFacilityImage(bitmap, facilityId)
            }
        }
        deferredUrls.awaitAll()
    }

    suspend fun uploadPitchImage(bitmap: Bitmap, facilityId: String, pitchId: String): String {
        val imageId = UUID.randomUUID().toString()
        val ref = pitchImagesRef.child(facilityId).child(pitchId).child("$imageId.jpg")
        return uploadImage(bitmap, ref)
    }

    suspend fun uploadPitchImages(bitmaps: List<Bitmap>, facilityId: String, pitchId: String): List<String> = coroutineScope {
        val deferredUrls = bitmaps.map { bitmap ->
            async {
                uploadPitchImage(bitmap, facilityId, pitchId)
            }
        }
        deferredUrls.awaitAll()
    }

    suspend fun uploadUserProfileImage(bitmap: Bitmap, userId: String): String {
        val ref = userImagesRef.child(userId).child("profile.jpg")
        return uploadImage(bitmap, ref)
    }

    suspend fun uploadVerificationDocument(uri: Uri, documentType: String): String {
        val userId = com.example.HaliSahaApp.data.remote.FirebaseService.currentUserId 
            ?: throw FirebaseError.NotAuthenticated
        val imageId = UUID.randomUUID().toString()
        val ref = adminVerificationsRef.child(userId).child("${documentType}_$imageId.jpg")
        
        // Actually, uploadImage requires a Bitmap. I'll just upload the Uri directly 
        // to avoid loading large images into memory unless necessary, or use putFile.
        return performUploadUri(uri, ref)
    }

    // MARK: - Core Upload & Optimization

    private suspend fun uploadImage(bitmap: Bitmap, ref: StorageReference): String {
        val imageData = optimizeImage(bitmap)
        
        if (imageData.size > MAX_IMAGE_SIZE) {
            throw FirebaseError.Unknown("Dosya boyutu çok büyük (max 5MB)")
        }

        return performUpload(imageData, ref)
    }

    private fun optimizeImage(bitmap: Bitmap): ByteArray {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()

        var ratio = minOf(MAX_IMAGE_DIMENSION / width, MAX_IMAGE_DIMENSION / height)
        if (ratio > 1.0f) ratio = 1.0f // Eğer resim 1920'den küçükse büyütme

        val newWidth = (width * ratio).toInt()
        val newHeight = (height * ratio).toInt()

        val resizedBitmap = Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
        
        val baos = ByteArrayOutputStream()
        resizedBitmap.compress(Bitmap.CompressFormat.JPEG, COMPRESSION_QUALITY, baos)
        return baos.toByteArray()
    }

    private suspend fun performUploadUri(uri: Uri, ref: StorageReference): String {
        var lastException: Exception? = null

        for (attempt in 1..MAX_RETRIES) {
            try {
                uploadMutex.withLock {
                    ref.putFile(uri).await()
                }
                val downloadUrl = ref.downloadUrl.await()
                return downloadUrl.toString()
            } catch (e: Exception) {
                lastException = e
                if (attempt < MAX_RETRIES) {
                    val delayTime = RETRY_DELAY_MS * (1 shl (attempt - 1)) // 1s, 2s, 4s
                    delay(delayTime)
                }
            }
        }
        throw lastException ?: FirebaseError.Unknown("Yükleme işlemi $MAX_RETRIES deneme sonrası başarısız oldu.")
    }

    private suspend fun performUpload(data: ByteArray, ref: StorageReference): String {
        val metadata = StorageMetadata.Builder()
            .setContentType("image/jpeg")
            .build()

        var lastException: Exception? = null

        for (attempt in 1..MAX_RETRIES) {
            try {
                uploadMutex.withLock {
                    ref.putBytes(data, metadata).await()
                }
                val downloadUrl = ref.downloadUrl.await()
                return downloadUrl.toString()
            } catch (e: Exception) {
                lastException = e
                if (attempt < MAX_RETRIES) {
                    val delayTime = RETRY_DELAY_MS * (1 shl (attempt - 1)) // 1s, 2s, 4s
                    delay(delayTime)
                }
            }
        }
        throw lastException ?: FirebaseError.Unknown("Yükleme işlemi $MAX_RETRIES deneme sonrası başarısız oldu.")
    }

    // MARK: - Deletion

    suspend fun deleteImage(url: String) {
        if (url.isEmpty()) return
        try {
            val ref = storage.getReferenceFromUrl(url)
            ref.delete().await()
        } catch (e: Exception) {
            // Hata loglanabilir
            println("Delete image failed for url: $url, error: ${e.localizedMessage}")
        }
    }

    suspend fun deleteImages(urls: List<String>) = coroutineScope {
        val deferredDeletes = urls.filter { it.isNotEmpty() }.map { url ->
            async {
                deleteImage(url)
            }
        }
        deferredDeletes.awaitAll()
    }

    suspend fun deleteFacilityImages(facilityId: String) {
        val ref = facilityImagesRef.child(facilityId)
        deleteFolder(ref)
    }

    suspend fun deletePitchImages(facilityId: String, pitchId: String) {
        val ref = pitchImagesRef.child(facilityId).child(pitchId)
        deleteFolder(ref)
    }

    private suspend fun deleteFolder(ref: StorageReference) {
        try {
            val result = ref.listAll().await()

            // Listedeki dosyaları paralel sil
            coroutineScope {
                result.items.map { item ->
                    async { item.delete().await() }
                }.awaitAll()
            }

            // Alt klasörleri dolaşıp sil
            for (prefix in result.prefixes) {
                deleteFolder(prefix)
            }
        } catch (e: Exception) {
            println("Delete folder failed: ${e.localizedMessage}")
        }
    }
}
